add partype method to lognormal and semicircle

bgctw opened the pull request:
Notice that LogNormal was missing the partype method. Hence, I added it and checked all uniform/continuous distributions. If they were parametric, I added the partype method for them.

devmotion commented on src/univariate/continuous/lognormal.jl:
```diff
@@ -62,6 +62,7 @@ stdlogx(d::LogNormal) = d.σ
 mean(d::LogNormal) = ((μ, σ) = params(d); exp(μ + σ^2/2))
 median(d::LogNormal) = exp(d.μ)
 mode(d::LogNormal) = ((μ, σ) = params(d); exp(μ - σ^2))
+partype(d::LogNormal{T}) where {T<:Real} = T
```
The argument name is not needed (in fact, the language server will complain about it):
```suggestion
partype(::LogNormal{T}) where {T<:Real} = T
```

devmotion commented on src/univariate/continuous/semicircle.jl:
```diff
@@ -34,6 +34,8 @@ Semicircle(r::Integer; check_args::Bool=true) = Semicircle(float(r); check_args=
 @distr_support Semicircle -d.r +d.r
 
 params(d::Semicircle) = (d.r,)
+partype(d::Semicircle{T}) where {T<:Real} = T
```
```suggestion
partype(::Semicircle{T}) where {T<:Real} = T
```

devmotion reviewed (commented):
It would be good to add a check for `partype` to the automatic distribution test suite (at least for univariate distributions), e.g., in https://github.com/JuliaStats/Distributions.jl/blob/2cd6c678a1f237462d71496c39178dab1e265c61/test/univariates.jl#L51. Something along the lines of `partype(d) === Base.promote_typeof(params...)` and `partype(d) === Float32` in the `Float32` conversion checks.

bgctw commented:
I implemented the tests for all distributions suggested by @devmotion. 

They cause the following problems.
- With TruncatedDistribution there may be Nothing types for some parameters (
  - took care of this special case in the checks (in ae09da6a32039c5e52e993bb515f3f8283e65f48)
- test/util.jl explitly tests partype of Hypergeometric and DiscreteUniform to be Float64, although only Integer-valued parameters are resonable for those distributions. 
  - I changed the tests (in b8511e9e53055422a50df9ccf57f829a1cf96a66) but this may brake code that depends on parameter types being fixed to Float64

devmotion commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ docs/site
 .Rhistory
 Manifest.toml
 .vscode
+tmp/
```
Revert?

devmotion replied:
```suggestion
```

devmotion commented on src/univariate/discrete/discreteuniform.jl:
```diff
@@ -21,21 +21,35 @@ External links
+struct DiscreteUniform{I} <: DiscreteUnivariateDistribution where I
+    a::I
+    b::I
     pv::Float64 # individual probabilities
 
-    function DiscreteUniform(a::Real, b::Real; check_args::Bool=true)
+    function DiscreteUniform{I}(a::I, b::I; check_args::Bool=true) where I <: Integer
         @check_args DiscreteUniform (a <= b)
-        new(a, b, 1 / (b - a + 1))
+        new{I}(a, b, one(I) / (b - a + one(I)))
+    end
+    function DiscreteUniform(a::Real, b::Real; kwargs...)
+        a_int = a isa Integer ? a : ceil(Int,a)
+        b_int = b isa Integer ? b : floor(Int, b)
+        aI, bI = promote(a_int,b_int)
+        I = typeof(aI)
+        DiscreteUniform{I}(aI,bI; kwargs...)
     end
     DiscreteUniform(b::Real; check_args::Bool=true) = DiscreteUniform(0, b; check_args=check_args)
-    DiscreteUniform() = new(0, 1, 0.5)
+    DiscreteUniform() = new{Int}(0, 1, 0.5)
 end
```
Can we avoid such major changes in the PR?

bgctw replied:
Originally, I introduced those modifications to discreteuniform to test how in addition to Float64/Float32 also Integer parameters could introduce a type parameter to allow using Float64/Float32. Since, this blows up the pull-request, I will move them to its own PR after this PR is hopefully accepted.

devmotion commented on src/univariate/discrete/discreteuniform.jl:
```diff
@@ -21,21 +21,35 @@ External links
 
 @distr_support DiscreteUniform d.a d.b
 
+partype(::DiscreteUniform{T}) where {T<:Integer} = T
```
Just
```suggestion
partype(::DiscreteUniform) = Int
```
?

devmotion commented on src/univariate/discrete/discreteuniform.jl:
```diff
@@ -21,21 +21,35 @@ External links
 
+partype(::DiscreteUniform{T}) where {T<:Integer} = T
+
+#### Conversions
+convert(::Type{DiscreteUniform{T}}, a::S, b::S) where {T <: Integer, S <: Real} = DiscreteUniform(T(a), T(b))
+Base.convert(::Type{DiscreteUniform{T}}, d::DiscreteUniform) where {T<:Integer} = DiscreteUniform{T}(T(d.a), T(d.b))
+Base.convert(::Type{DiscreteUniform{T}}, d::DiscreteUniform{T}) where {T<:Integer} = d
+
```
Revert?
```suggestion
```

bgctw replied:
Will adapt this when reverting the DiscreteUniform to non-parametric.

devmotion commented on src/univariate/discrete/discreteuniform.jl:
```diff
@@ -73,7 +87,7 @@ modes(d::DiscreteUniform) = [d.a:d.b]
 logpdf(d::DiscreteUniform, x::Real) = log(pdf(d, x))
 
-function cdf(d::DiscreteUniform, x::Int)
+function cdf(d::DiscreteUniform, x::Integer)
```
Unrelated?
```suggestion
function cdf(d::DiscreteUniform, x::Int)
```

bgctw replied:
Will adapt this when reverting the DiscreteUniform to non-parametric.

devmotion commented on src/univariate/discrete/discreteuniform.jl:
```diff
@@ -85,7 +99,10 @@ function cdf(d::DiscreteUniform, x::Int)
 end
 
-quantile(d::DiscreteUniform, p::Real) = iszero(p) ? d.a : d.a - 1 + ceil(Int, p * span(d))
+function quantile(d::DiscreteUniform, p::Real) 
+    T = partype(d)
+    iszero(p) ? d.a : d.a - one(T) + ceil(T, p * span(d))
+end
```
Revert since unrelated?

bgctw replied:
Will adapt this when reverting the DiscreteUniform to non-parametric.

devmotion commented on src/univariate/locationscale.jl:
```diff
@@ -49,10 +49,15 @@ struct AffineDistribution{T<:Real, S<:ValueSupport, D<:UnivariateDistribution{S}
 
-function AffineDistribution(μ::T, σ::T, ρ::UnivariateDistribution; check_args::Bool=true) where {T<:Real}
+
+function AffineDistribution(μ::T, σ::T, ρ::D; check_args::Bool=true) where {T<:Real,D<:UnivariateDistribution}
```
```suggestion
function AffineDistribution(μ::T, σ::T, ρ::UnivariateDistribution; check_args::Bool=true) where {T<:Real}
```

devmotion commented on test/univariates.jl:
```diff
@@ -62,6 +62,13 @@ function verify_and_test(D::Union{Type,Function}, d::UnivariateDistribution, dct
     pars = params(d)
 
+    # verify parameter type
+    # truncated parameters may be nothing: Union{Nothing, promote_type()}
+    # but partype should still be type of the non-nothing ones
+    #@test partype(d) === promote_type(typeof.(pars)...)
+    @test partype(d) === promote_type(typeof.(pars)...) || 
+      Union{Nothing,partype(d)} ===  promote_type(typeof.(pars)...)
```
Maybe just use something like
```suggestion
    # truncated parameters may be nothing: Union{Nothing, promote_type()}
    # but partype should still be type of the non-nothing ones
    @test partype(d) === mapfoldl(typeof, (S, T) -> T <: Real ? promote_type(S, T) : S, pars; init = Bool)
```

bgctw replied:
Had to wrap my head around, but implemented and tested this more general implementation with an own commit.

devmotion commented on test/univariates.jl:
```diff
@@ -75,10 +82,18 @@ function verify_and_test(D::Union{Type,Function}, d::UnivariateDistribution, dct
         @test convert(D{partype(d)}, d) === d
-        d32 = convert(D{Float32}, d)
-        @test d32 isa D{Float32}
+        if D <: DiscreteUniform #|| D  <: Hypergeometric
+            d32 = convert(D{Int32}, d)
+            @test d32 isa D{Int32}
+            @test partype(d32) == Int32
+        else
+            d32 = convert(D{Float32}, d)
+            @test d32 isa D{Float32}
+            @test partype(d32) == Float32
+        end
```
If you remove the type parameter from `DiscreteUniform` again, you should not be able to enter this branch anymore (and hence do not need to handle this case here) since `isconcretetype(DiscreteUniform) === true`.
```suggestion
        d32 = convert(D{Float32}, d)
        @test d32 isa D{Float32}
        @test partype(d32) == Float32
```

bgctw replied:
Modifying DiscreteUniform was intentend to support this test-case of a parametric Integer type. I will move it to its own pull request.

bgctw commented:
Based on the code comments of @devmotion, I reverted the changes to discreteuniform. 

Originally, I introduced those modifications to discreteuniform to test how in addition to Float64/Float32  Distribution can also use a type parameter to allow using Int64/Int32. Since, this blows up the pull-request, I will create an PR after this PR is hopefully accepted.

bgctw commented:
Do I still need to do something to move this PR forward?

devmotion commented on test/univariates.jl:
```diff
@@ -62,6 +62,12 @@ function verify_and_test(D::Union{Type,Function}, d::UnivariateDistribution, dct
+    # truncated parameters may be nothing: Union{Nothing, promote_type()}
+    # in the following exclude all non-Real from creating the promoted type
+    @test partype(d) === mapfoldl(
+        typeof, (S, T) -> T <: Real ? promote_type(S, T) : S, pars; init = Bool)
```
```suggestion
        typeof, (S, T) -> T <: Distribution ? promote_type(S, partype(T)) : (T <: Nothing ? S : promote_type(S, eltype(T))), pars; init = Union{})
```

devmotion commented on test/utils.jl:
```diff
@@ -15,8 +15,14 @@ r = RealInterval(1.5, 4.0)
 @test partype(Kolmogorov()) == Float64
-@test partype(Hypergeometric(2, 2, 2)) == Float64
-@test partype(DiscreteUniform(0, 4)) == Float64
+# twutz: Hypergeometric needs integer parameters
+#@test partype(Hypergeometric(2, 2, 2)) == Float64
```
```suggestion
```

devmotion commented on test/utils.jl:
```diff
@@ -15,8 +15,14 @@ r = RealInterval(1.5, 4.0)
+#@test partype(Hypergeometric(2, 2, 2)) == Float64
+@test partype(Hypergeometric(2, 2, 2)) == Int
+@test partype(Hypergeometric(2.0, 2, 2)) == Int
+# twutz: should DiscreteUniform needs a partype of Int
+#@test partype(DiscreteUniform(0, 4)) == Float64
```
```suggestion
```

devmotion reviewed (approved):
Thank you!
